lib/AST/DeclContext.cpp

```diff
@@ -258,6 +258,23 @@ DeclContext *DeclContext::getParentForLookup() const {
   return getParent();
 }
 
+SourceLoc DeclContext::getLocForLookupInParent(SourceLoc loc) const {
+  if (auto *D = getAsDecl()) {
+    if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
+      // (In code completions,) the body of the AFD might be parsed from another
+      // source buffer. To avoid comparing SourceLocs between different buffers,
+      // use the location of the decl itself for lookup in parent context.
```

Comment:
Nice comment, and nicely done!

Reply:
You know, I realized that I still don't understand how this can work? Why supply a phoney-baloney location rather than just fix the location comparison code in lookup?
Is this location used for something? Is something at this location?

Reply:
Would an invalid location work just as well??

Reply:
Is this about inner functions??

Reply:
> Why supply a phoney-baloney location rather than just fix the location comparison code in lookup?

Good question. Changing the location comparison code is probably a "correct" fix. But I believe it's too heavy. `SourceManager` have to remember that some portion of the buffer is replaced by another file, and comparison code will be slower (compared to the current pointer comparison).

> Is this location used for something? Is something at this location?

No and No. But all we need here is to propagate a source location which is inside the parent context. Since we already processed the AFD itself before walk to the parent context, we don't need to preserve the exact location inside the AFD.

> Would an invalid location work just as well??

Are you asking what if `AbstractFunctionDecl::getLoc()` returns `getLoc()`? Or what if the original location is invalid?

Reply:
I do like the idea of changing the location comparison code, and look forward to exploring that with you so I can understand it better.

If the location is not used for anything, then supplying a fake one may be confusing.

Yes, either or both questions: a cursory inspection suggests that the lookup code is intended to handle invalid locations. Simplest for it would be to pass in an original location that is invalid, perhaps. But I look forward to gaining a better understanding of completion from you to understand the use-case better.